server/models/docker_image.py

```diff
@@ -0,0 +1,279 @@
+            spec_dict[key] = val[DockerImage.type]
+        return spec_dict
+
+    def getRawData(self):
+        return self.data
+
+
+class DockerCache:
+    """
+    This class is used to hold and access meta data on existing images
+    """
+    def __init__(self):
+        """
+        Data is stored in the following format:
+            {image_name_hash:DockerImage
+
+            }
+        """
+
+        self.data = {}
+
+    def addImage(self, img):
+        """
+        Add an image object to the Docker cache
+        :param img: A docker image object
+        :type img: DockerImage
+        """
+        try:
+            if isinstance(img, DockerImage):
+                self.data[img.hash] = DockerImage(img.getRawData())
+            else:
+                raise DockerImageError('Tried to add a non '
+                                       'docker image object to cache')
+        except Exception as err:
+            raise DockerImageError('Failed to add the '
+                                   'docker image to the cache' + err.__str__())
+
+    def getDockerImageList(self):
+        """
+        Get the list docker image names in the cache
+        """
+        return [img.name
+                for (imgHash, img) in iteritems(self.data)]
+
+    def getImages(self):
+        """
+        Get a list of Docker images objects stored in the cache
+        """
+        return [img
+                for (imgHash, img) in iteritems(self.data)]
+
+    def getImage(self, name):
+        """
+        Get an image object using the Docker image name
+        :param name: The docker image name
+        :type name:string
+        """
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            return self.data[imageKey]
+
+    def isImageAlreadyLoaded(self, name):
+        """
+
+        Checks whether an image was already loaded, via docker image name.
+        This check does not use the docker image id, and therefore will not
+        treat two equivalent images with different names as similar
+        :param name: The docker image name
+        :type name:string
+        """
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            return True
+        else:
+            return False
+
+    def _getHashKey(self, name):
+        return DockerImage.getHashKey(name)
+
+    def getRawData(self):
+        dataCopy = {}
+        for (key, val) in iteritems(self.data):
+            dataCopy[key] = val.getRawData()
+        return dataCopy
+
+    def deleteImage(self, name):
+        imageKey = self._getHashKey(name)
+        if imageKey in self.data:
+            del self.data[imageKey]
+            return True
+        else:
+            return False
+
+    def getAllCliSpec(self):
+        spec_dict = {}
+
+        for (key, val) in iteritems(self.data):
+            spec_dict[val.name] = val.getCLIListSpec()
+
+        return spec_dict
```

Comment:
The code from line 200 to here is largely uncovered by the tests.  If it isn't too much work, I'd love to see a test for it.